rust/lance-file/src/datatypes.rs

```diff
@@ -0,0 +1,310 @@
+use std::collections::HashMap;
+
+use arrow_schema::DataType;
+use async_recursion::async_recursion;
+use lance_arrow::bfloat16::ARROW_EXT_NAME_KEY;
+use lance_core::datatypes::{Dictionary, Encoding, Field, LogicalType, Schema};
+use lance_core::{Error, Result};
+use lance_io::traits::Reader;
+use lance_io::utils::{read_binary_array, read_fixed_stride_array};
+use snafu::{location, Location};
+
+use crate::format::pb;
+
+impl From<&pb::Field> for Field {
+    fn from(field: &pb::Field) -> Self {
+        let mut lance_metadata: HashMap<String, String> = field
+            .metadata
+            .iter()
+            .map(|(key, value)| {
+                let string_value = String::from_utf8_lossy(value).to_string();
+                (key.clone(), string_value)
+            })
+            .collect();
+        if !field.extension_name.is_empty() {
+            lance_metadata.insert(ARROW_EXT_NAME_KEY.to_string(), field.extension_name.clone());
+        }
+        Self {
+            name: field.name.clone(),
+            id: field.id,
+            parent_id: field.parent_id,
+            logical_type: LogicalType::from(field.logical_type.as_str()),
+            metadata: lance_metadata,
+            encoding: match field.encoding {
+                1 => Some(Encoding::Plain),
+                2 => Some(Encoding::VarBinary),
+                3 => Some(Encoding::Dictionary),
+                4 => Some(Encoding::RLE),
+                _ => None,
+            },
+            nullable: field.nullable,
+            children: vec![],
+            dictionary: field.dictionary.as_ref().map(Dictionary::from),
+        }
+    }
+}
+
+impl From<&Field> for pb::Field {
+    fn from(field: &Field) -> Self {
+        let pb_metadata = field
+            .metadata
+            .clone()
+            .into_iter()
+            .map(|(key, value)| (key, value.into_bytes()))
+            .collect();
+        Self {
+            id: field.id,
+            parent_id: field.parent_id,
+            name: field.name.clone(),
+            logical_type: field.logical_type.to_string(),
+            encoding: match field.encoding {
+                Some(Encoding::Plain) => 1,
+                Some(Encoding::VarBinary) => 2,
+                Some(Encoding::Dictionary) => 3,
+                Some(Encoding::RLE) => 4,
+                _ => 0,
+            },
+            nullable: field.nullable,
+            dictionary: field.dictionary.as_ref().map(pb::Dictionary::from),
+            metadata: pb_metadata,
+            extension_name: field
+                .extension_name()
+                .map(|name| name.to_owned())
+                .unwrap_or_default(),
+            r#type: 0,
+        }
+    }
+}
+
+pub struct Fields(pub Vec<pb::Field>);
```

Comment:
Minor introduction here of `Fields` and (later) `FieldsWithMeta`.  This is because `Field` and `Schema` are in `lance-core` but the to/from protobuf methods are in `lance-file`.  You can't do `impl From<X> for Vec<Y>` if `X` doesn't belong to your crate (even if `Y` does).

Reply:
I guess it makes some sense to put this here as it is the deserialization part, but part of me still thinks it's easier to just keep this associated with the `Schema` definition. We could consider having a separate `schema.proto`.

Reply:
I think making a `schema.proto` is fine (can do later).  I agree it is a little inconsistent.